### ISSUE TITLE
UWP App Working on Lumia 950XL (ARM)

### ISSUE THREAD

**Issue by kburgoyne** [sharpdx/SharpDX#924]
Just FYI... After migrating from a Win8 Universal app using SharpDX v3 to a Win10 UWP app using SharpDX v4 I've just completed running for the first time a release build on a Lumia 950XL (ARM).  This means the Native Tool Chain targeting ARM.  No apparent issues.  I've also previously run Win10 UWP x86 release builds using the Native Tool Chain targeting x86.

Compliments to the SharpDX Chefs.  :-)

**Comment by xoofx** [comment]
Good, thanks for the feedback!

**Comment by JimSEOW** [comment]
@kburgoyne could you share a simple example running on 950XL (ARM). It would be great for WinOnARM Q4 2017

**Comment by kburgoyne** [comment]
The best thing would be just a UWP app version of the tumbling cube without all the fancy code structuring used in the SharpDX v3 samples.  The v3 samples provide a lot of variation in what they show, but they're not optimal for somebody just getting started and just wants something very simple that is easily understood.

Is there a Microsoft page/posting/etc about "WinOnArm Q4 2017"?  There is sooo much going on with Azure, UWP, ASP, IoT, etc, etc, etc, that I can't keep up and actually get any work done at the same time.

All I can say is I'll put a cube demo on my "to do" list and try to find a breathing point to create one.  I have some milestones I'm trying to hit right now.  I keep thinking "I'm almost there", but then little supporting tasks keep cropping up.

@xoofx  -- Does SharpDX have any kind of logo associated with it? If I find time to create a UWP tumbling cube I'll start with solid colors to keep it the simplest to follow. I could look at a second version which paints a texture on the cube. Once the solid colors is done it's a minor step to add the texture.

**Comment by JimSEOW** [comment]
@kburgoyne  @xoofx  There are now DirectX11 and Direx12 interop XAML templates. It make sense to have similar C# templates so users who have done many excellent SharpDX11 and SharpDX12 in console can just port that to XAML UWP. 

I have seen people shared one through the one Drive but then delete it. 

I wonder why people are so worry to share this standard c# ported version of the templates?

Is SharpDX not stable for these purpose???

**Comment by JimSEOW** [comment]
Another [reason ](https://github.com/Microsoft/WindowsTemplateStudio/issues/561#issue-240172652)to make SharpDX UWP XAML (SwapchainPanel) ready

**Comment by Gavin-Williams** [comment]
I don't understand this thread, SharpDX already works with SwapChainPanel.

**Comment by JimSEOW** [comment]
@Gavin-Williams there are many challenges is using SharpDX for UWP in C# XAML interop for new users
(a) there is only One working UWP .net XAML interop sample that is based on AND ported from C++ DirectX11 templates. The public sample sharpdx store sample based on 8.1 is 2 years old. Everyone told everyone new user to F#%$# himself to use it for UWP win10 XAML interop .net. It should be straight forwards. 

Person who like @kburgoyne who asked many many questions since early May how to do UWP Xaml interop. And has to go through these F#$%## .

I recently read these Feedback from him and thought it was easy to do. And gave it a try to discover these F#%$##

Shapdx is a trap for UWP Win10 XAML interop user to waste time

**Comment by xoofx** [comment]
@JimSEOW  What do you mean is a trap? The complexity doesn't come from SharpDX, but DirectX and UWP interop. 

DirectX is not straightforward, interop with XAML Windows 10 the same, so don't expect any added values for SharpDX, as it is just a low level wrapper of these technologies. You have to go through the existing DirectX C++ documentation and code samples and translate them to C#. But If you are looking for a higher level interaction, SharpDX is definitely not the right library (use MonoGame, Unity, Xenko...etc.)

Typically, how do you think I ported the [Store samples](https://github.com/sharpdx/SharpDX-Samples/tree/master/StoreApp)? I just took the original C++ sample code and ported them straight to C#. I didn't have to face any hurdle as it was a plain translation. So now, I did port these samples 2 years ago, it took me quite some time to do them (and frankly, this is very boring)... 

Notice that since then, nobody helped there to bring any new samples... 

So as I'm a bit "done" with SharpDX today (no longer using it on my own or in my company), I'm no longer able to provide this huge work - for free- for the community, but I will gladly integrate any PR that can bring new samples.

**Comment by JimSEOW** [comment]
@xoofx, You have done a good job to make SharpDX reality. New users trying uwp for XAML interop for Win10 has to go through SAME mistakes and not sharing them.

I just want to share the challenges I went through so others have a clear picture to estimate the challenges they have to go through

**Comment by JimSEOW** [comment]
@xoofx  There is ALREADY [one working UWP Win10 XAML Interop sample](https://github.com/minhcly/UWP3DTest/issues/1#issue-240043163) and I suggest we put that to the [Official Sample site.](https://github.com/sharpdx/SharpDX-Samples/tree/master/StoreApp)

**Comment by kburgoyne** [comment]
Considering nobody gets paid to work on SharpDX, the tone of the comments seems out of line.  Actually writing a beginners SharpDX example for future beginners to reference might prove to be an excellent place to start for a beginner.

The bottom line is, SharpDX is DirectX.  Those who have mastered DirectX using C++ had a very steep learning curve as well.  DirectX is very close to being "down on the metal" of the GPU itself.  The basics of how GPUs go about their business is pretty exposed at the DirectX level.  Like with DirectX, packages which abstract stuff even further from the GPU level and thus make things easier for developers are packages which sit on top of DirectX and are not part of DirectX itself.  Thus the same would be true for SharpDX, which is really just DirectX.

**Comment by JimSEOW** [comment]
@kburgoyne NO ONE IS ASKING YOU TO WRITE !!!!   But I read every single feedback.  I see the pain you have gone through

**Comment by JimSEOW** [comment]
In most universities, if you joined as a fresh man, you have to go through the shit orientation so that your senior one year older can shit on you. 

Same thing is happening here I just want ALL NEW USERS to UWP Win10 XAML interop (Not SharpDX beginner who already finished porting DirectX11 book samples to SharpDX) to have a clear pictures of the challenges ahead and decide how to contribute

**Comment by JimSEOW** [comment]
@xoofx We now have a great opportunity to use SharpDX to contribute to education of the early age children for a higher purpose. Thanks for your 7 years of Contribution so new SharpDX users like myself can continue your journey to contribute.

**Comment by xoofx** [comment]
So as I said, PR are welcome.

**Comment by JimSEOW** [comment]
Next things to do for SharpDX Win10 XAML interop is to gather Systematic feedback from users on the new insider Win10 Builds.

In what ways the latest SharpDX codes failed with the latest Win10 insider builds (in my experience 10.0.15063 and perhaps beyond)

**Comment by kburgoyne** [comment]
The one thing from your feedback that I'm seeing is the desire not just for a working UWP sample, but preferably a UWP sample that also serves as a kind of tutorial.  I did start a little work on that.  However the fact remains that ultimately it's DirectX.

Such a tutorial can expose people to the "right way of thinking" about what DirectX really is.  I do remember when I had all the wrong notions about what DirectX is.  However I have an ancient background writing assembly language and even some machine language, so I had a foundation upon which to start to grasp how DirectX was a reflection of the GPU architecture.  (And probably visa-versa as GPUs and DirectX evolved over time.)

In the end, beginners are still faced with it being DirectX which is very low-level.  Most likely one of the largest benefits of such a tutorial would be to puncture people's misperceptions (like I had) about what DirectX is and what it isn't.

It is true such a tutorial could answer a whole lot of questions about getting DirectX initialized, the need for vertex buffers, etc, and perhaps even some early exposure to "the general concept" of how vertex buffers feed into vertex and pixel shaders.  A tutorial can even provide some very primitive shaders to provide some exposure to the concept.  So I guess that's constructive.  I probably would have benefitted.

The warning is that once the basic concepts are understood, that just opens the door to even more complexity.  However certainly grasping that complexity is easier once one gets an understanding of the basic DirectX concepts.

Regarding your comment about 7 year olds.  They're not going to be using DirectX/SharpDX directly unless they're some kind of wizard.  It's an admirable idea, but so is teaching jet aircraft design to 7 years olds.  It's not an introductory subject level. Far far from it.  I wouldn't dismiss a 7 year old wizard as being unable to figure it all out, but that's not "introductory".  That's a 7 year old geek digging deep.  (Keep track of him and hire him 10 years later.)

What keeps being emphasized is that, for introductory level, somebody would have to write "something" which sits on top of SharpDX and hides (takes care of) most of the complexity of DirectX.  What approach that "something" takes would be best guided by somebody who actually works with the kids.  Certainly the Remix3D stuff has positives in that regard.

I believe I asked above. Do you have any links to pages which discuss Remix3D from a more developer standpoint?  Or at least the details that a developer would need to know?

**Comment by JimSEOW** [comment]
@kburgoyne  Thanks for sharing your thought. I respect your persistence in getting UWP XAML interop for SharpDX right. 

I tend to see thing in a bigger picture. 

The bigger picture, right now, DirectX has always belonged to C++ elite group. It is great that we have Managed DirectX, SlimDX and NOW the keeping up-to-date ShapeDX for .NET community.

In the coming competitive landscape of Microsoft Mixed Reality, due to challenge by Apple ARkit. I know HoloLens are miles ahead of the technology. However, I see Customer Centricity. What need to inspire a child to imagine with no limitation, all this child need is to have either a iPhone and iPAD to express his or her world through creative superimposition on live stream camera.  There is no need for Top HoloLens Technologies here  We just need to empower the "Good enough tool" to LET US See their world.

Apple has empowered its developers with these tools OVERNIGHT an AR standard.

Microsoft still reply ON TOP Technology excellent to come to consumer "HoloLens" - glass (e.g. 2019). Even after that, not many can afford. For Apple, all u need is a traveller, who bring an iPhone or IPad, perhaps to a remote corner in Africa, let children go crazy with this. The impact at this moment is priceless. 

Unfortunately, Microsoft Engineers still argue Technical Excellence over spreading this JOY to Children World Wide NOW, not 2019.

Microsoft pushing Mixed Reality really hard. At the same time, there is no CONNECTION between Story 3D remix and Mixed Reality headset. This shocks ME !!!!!

Ideally, Story 3D Remix will be available to WinOnARM Q4 2017. Instead of superimpose keyframe 3D animation on recorded video, there should be an easy add on feature of Live Stream Video.

Still, Microsoft Should Open out the Story 3D Remix API **for developers like yourself** to stretch your imagination and creativity. 

I therefore using Microsoft/WindowsTemplateStudio as a stepping stone to make Microsoft aware What (we .NET SharpDX or OpenGLES developers) need.

I hope they see the merit to eventually open out more API information on Story 3D Remix. 

I hope they are aware how Software Mixed Reality through AR on top of live stream video is also strategic to the Microsoft AI driven education global plan, especially for early age children. 

In summary, I have been knocking on different ways to make them aware. I appreciate your interest to do something about it TOO.

**Comment by Gavin-Williams** [comment]
@JimSEOW  Not related to SharpDX ... 

 "In most universities, if you joined as a fresh man, you have to go through the shit orientation so that your senior one year older can shit on you."

What universities do this ? I've been to three different universities (Australia), and that's not my experience at all. First year university is amazing, and very supportive. No one ever shit on me. Students absolutely love orientation, all Australian universities have a tradition of orientation being rich with activities and entertainment etc. I can't imagine other universities around the world not doing something similar or especially making their first year students have a 'shit time'. Have you been to university yourself? Because it doesn't sound like it.

**Comment by JimSEOW** [comment]
@Gavin-Williams You seems to be a reasonably smart person AND curious to be able to google to investigate how often the incidence I claim is happening BEYOND your University experience to others.

Still, Let us go back to SharpDX.
